### PR TITLE
Add authentication to API and retrieve headers from querying graph database

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

Add authentication to private endpoints using dp-authorization library. Also retrieve headers for calculating csv dimension offset for finding observation metadata as well as:

- fix self link to use observation API url instead of using the dataset API url retrieved from version document.
- add zebedee checker to healthcheck for publishing instance
- handle errors from dataset API
- update unit tests
- update dependencies to use latest library versions, also use dp-net instead of deprecated dp-rchttp

### How to review

- Check changes make sense and you can successfully retrieve an observation.
- Check unit tests pass.

### Who can review

Anyone

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -4,39 +4,58 @@ import (
...
+	}
+
+	userIdentity := common.User(r.Context())
+	if userIdentity != "" {
+		logData["user_identity"] = userIdentity
+		authorised = true
 	}
+
+	logData["authenticated"] = authorised
+
```
Comment (CarlHembrough) [inline api/api.go:58]: This looks like some common logic to check if the request is authorised - could this be put into a library for reuse?

File: api/observation.go [hunk 2]
```diff
@@ -93,23 +95,23 @@ func (api *API) doGetObservations(ctx context.Context, datasetID, edition, versi
 		return nil, err
 	}
 
-	if versionDoc.CSVHeader == nil || versionDoc.Dimensions == nil {
+	if err = models.CheckState(models.Version, versionDoc.State); err != nil {
+		logData["state"] = versionDoc.State
+		log.Event(ctx, "get observations: unpublished version has an invalid state", log.ERROR, log.Error(err), logData)
```
Comment (CarlHembrough) [inline api/observation.go]: Will it always be an unpublished version here?

File: api/observation.go [hunk 3]
```diff
@@ -93,23 +95,23 @@ func (api *API) doGetObservations(ctx context.Context, datasetID, edition, versi
...
+		return nil, err
+	}
+
+	if versionDoc.Dimensions == nil {
 		logData["version_doc"] = versionDoc
-		log.Event(ctx, "get observations", log.ERROR, log.Error(errs.ErrMissingVersionHeadersOrDimensions), logData)
-		return nil, errs.ErrMissingVersionHeadersOrDimensions
+		log.Event(ctx, "get observations", log.ERROR, log.Error(errs.ErrMissingVersionDimensions), logData)
```
Comment (CarlHembrough) [inline api/observation.go]: should there be some additional context in this log message? (maybe the error type provides enough context)

File: go.mod [hunk 4]
```diff
@@ -3,15 +3,16 @@ module github.com/ONSdigital/dp-observation-api
 go 1.13
 
 require (
-	github.com/ONSdigital/dp-api-clients-go v1.9.0
-	github.com/ONSdigital/dp-graph/v2 v2.0.3
-	github.com/ONSdigital/dp-healthcheck v1.0.3
-	github.com/ONSdigital/dp-rchttp v1.0.0 // indirect
-	github.com/ONSdigital/go-ns v0.0.0-20200205115900-a11716f93bad
+	github.com/ONSdigital/dp-api-clients-go v1.10.1
+	github.com/ONSdigital/dp-authorisation v0.0.0-20200416132558-8ad7b7fd9cee
```
Comment (CarlHembrough) [inline go.mod]: Do we need a release for this before merging?

**Review by CarlHembrough — CHANGES_REQUESTED** [review]
Looking good 👍 - just some minor points to consider

File: go.mod (hunk 4 above)
Comment (nshumoogum) [reply go.mod]: This is the first application to try and use dp-authorisation library, I will create a v0.1.0 version of lib and update go mod here

File: api/api.go (hunk 1 above)
Comment (nshumoogum) [reply api/api.go:58]: I think you are right, probably want to add it to `dp-authorisation` library; again I would look to do this once we re-use this logic in other applications (sessions/recipe API).

**Review by CarlHembrough — APPROVED** [review]

**Pull request merged**